go/mysqlconn/server.go

```diff
@@ -168,72 +168,53 @@ func (l *Listener) handle(conn net.Conn, connectionID uint32) {
-		if !l.authServer.UseClearText() {
-			userData, err := l.authServer.ValidateHash(salt, user, authResponse)
-			if err != nil {
-				c.writeErrorPacketFromError(err)
-				return
-			}
-			c.User = user
-			c.UserData = userData
-			// We're good.
-			break
-		}
+	// See what auth method the AuthServer wants to use for that user.
+	authServerMethod, err := l.authServer.AuthMethod(user)
+	if err != nil {
+		c.writeErrorPacketFromError(err)
+		return
+	}
 
-		// Our AuthServer cannot use mysql_native_password, it
-		// needs the real password.  Let's request that.
-		renegotiateWithClearText = true
-	case mysqlClearPassword:
-		// Client sent us a clear text password. Let's use it if we can.
-		if !l.AllowClearTextWithoutTLS && c.Capabilities&CapabilityClientSSL == 0 {
-			c.writeErrorPacket(CRServerHandshakeErr, SSUnknownSQLState, "Cannot use clear text authentication over non-SSL connections.")
-			return
-		}
-		userData, err := l.authServer.ValidateClearText(user, string(authResponse))
+	// Compare with what the client sent back.
+	switch {
+	case authServerMethod == MysqlNativePassword && authMethod == MysqlNativePassword:
+		// Both server and client want to use MysqlNativePassword:
+		// the negotiation can be completed right away, using the
+		// ValidateHash() method.
+		userData, err := l.authServer.ValidateHash(salt, user, authResponse)
 		if err != nil {
 			c.writeErrorPacketFromError(err)
 			return
 		}
 		c.User = user
 		c.UserData = userData
-		break
+
+	case authServerMethod == MysqlNativePassword:
+		// The server really wants to use MysqlNativePassword,
+		// but the client returned a result for something else:
+		// not sure this can happen, so not supporting this now.
+		c.writeErrorPacket(CRServerHandshakeErr, SSUnknownSQLState, "Client asked for auth %v, but server wants auth mysql_native_password", authMethod)
+		return
+
 	default:
-		// Client decided to use something we don't understand.
-		// Let's try again with clear text password.
-		renegotiateWithClearText = true
-	}
+		// The server wants to use something else, re-negotiate.
 
-	// If we need to re-negotiate with clear text, do it.
-	if renegotiateWithClearText {
-		// Check error conditions.
+		// The negotiation happens in clear text. Let's check we can.
 		if !l.AllowClearTextWithoutTLS && c.Capabilities&CapabilityClientSSL == 0 {
 			c.writeErrorPacket(CRServerHandshakeErr, SSUnknownSQLState, "Cannot use clear text authentication over non-SSL connections.")
 			return
 		}
 
-		if err := c.writeAuthSwitchRequest(mysqlClearPassword, nil); err != nil {
+		// Switch our auth method to what the server wants.
+		// Note: for now don't support the extra data.
+		if err := c.writeAuthSwitchRequest(authServerMethod, nil); err != nil {
 			log.Errorf("Error write auth switch packet for client %v: %v", c.ConnectionID, err)
 			return
 		}
 
-		// The client is supposed to just send the data in a single packet.
-		// It is a zero-terminated string.
```

Comment:
I did a bunch more testing tonight and have come up with a couple takeaways, though i feel like i'm not that much closer to figuring it out:

- I comment here because I notice this code is missing in the new rev. `writeAuthSwitchRequest` should trigger a AuthSwitchResponse, according to the mysql protocol docs. I don't see us reading that.  I tried hacking this code back in in the dialog path, and it didn't help -- failed on the same error.

The error is coming from ReadFully, as it suggests. It happens when trying to read the response from the client after sending the prompt. The return result is 0 bytes long. I think the client has not sent anything yet.

Looking at https://github.com/percona/percona-server/blob/1e2f003a5bd48763c27e37542d97cd8f59d98eaa/plugin/percona-pam-for-mysql/src/dialog.c#L312 I think we should be sending `0x4` as the first byte in the prompt question. I tried adding that in quickly and it didn't seem to help much though. Will try to do some more testing tomorrow.

Another thing, I moved the dialog.so files out of the way so they should theoretically not be usable by the client. But I'm not getting the usual "dialog.so file not found" error that you typically get in this case, so I almost wonder if the writeAuthSwitchRequest is working at all.